src/cargo/ops/cargo_report/timings.rs

```diff
@@ -99,3 +153,228 @@ fn select_log_file(
+        unit_data: Default::default(),
+        concurrency: Default::default(),
+        cpu_usage: Default::default(),
+        rustc_version: Default::default(),
+        host: Default::default(),
+        requested_targets: Default::default(),
+        jobs: 0,
+        error: &None,
+    };
+    let mut units: IndexMap<_, UnitEntry> = IndexMap::new();
+
+    for (log_index, result) in serde_json::Deserializer::from_reader(reader)
+        .into_iter::<LogMessage>()
+        .enumerate()
+    {
+        let msg = match result {
+            Ok(msg) => msg,
+            Err(e) => {
+                tracing::warn!("failed to parse log message at index {log_index}: {e}");
+                continue;
+            }
+        };
+
+        match msg {
+            LogMessage::BuildStarted {
+                cwd: _,
+                host,
+                jobs,
+                profile,
+                rustc_version,
+                rustc_version_verbose: _,
+                target_dir: _,
+                workspace_root: _,
+            } => {
+                ctx.host = host;
+                ctx.jobs = jobs;
+                ctx.profile = profile;
+                ctx.rustc_version = rustc_version;
+            }
+            LogMessage::UnitStarted {
+                package_id,
+                target,
+                mode,
+                index,
+                elapsed,
+            } => {
+                let version = package_id
+                    .version()
+                    .map(|v| v.to_string())
+                    .unwrap_or_else(|| "N/A".into());
+
+                // This is pretty similar to how the current `core::compiler::timings`
+                // renders `core::manifest::Target`. However, our target is
+                // a simplified type so we cannot reuse the same logic here.
+                let mut target_str = if target.kind == "lib" && mode == CompileMode::Build {
+                    // Special case for brevity, since most dependencies hit this path.
+                    "".to_string()
+                } else if target.kind == "build-script" {
+                    " build-script".to_string()
+                } else {
+                    format!(r#" {} "{}""#, target.name, target.kind)
+                };
+
+                match mode {
+                    CompileMode::Test => target_str.push_str(" (test)"),
+                    CompileMode::Build => {}
+                    CompileMode::Check { test: true } => target_str.push_str(" (check-test)"),
+                    CompileMode::Check { test: false } => target_str.push_str(" (check)"),
+                    CompileMode::Doc { .. } => target_str.push_str(" (doc)"),
+                    CompileMode::Doctest => target_str.push_str(" (doc test)"),
+                    CompileMode::Docscrape => target_str.push_str(" (doc scrape)"),
+                    CompileMode::RunCustomBuild => target_str.push_str(" (run)"),
+                }
```

Comment:
We probably want to reconsider using the old `cargo::core::Target` struct instead of the new simplified one.

Or better, HTML from `--timings` should be generated by in-memory log message, so that we don't need to have two parallel implementations doing the same thing.